### DISCUSSION TITLE
proposal: filter button panel

### DISCUSSION THREAD

**Discussion by quickfold** [freeplane/freeplane#1993] [Ideas]
I have an idea that would significantly improve Freeplane's usability for me, maybe for others also. The least fluid part of using Freeplane for me is filters. Filters are currently applied via dropdown menus and a lot of menu buttons. All have distinct functions, but it's hard to remember what they all do. I have to read the tooltips much of the time. Filters are edited in a separate compose window that allows very powerful combinations of filters to define specific groups. Even though filters are powerful, they are not convenient to use when you want to switch between filters quickly and often, especially when your use case is usually a simple one, such as "show only nodes with style 'Topic'". The compose window allows very precise filter combinations, but it requires you to leave the map and think in a very different (non-visual) way, and you don't see the results immediately.

I would like to propose the addition of a filter panel that is similar to Zotero's tag panel. It would allow a use of filters that is much faster and lets you figure out the right filter combination by exploration rather than having to do mental processing of logical categories. The filter panel (it could also be a tool pane tab) would show filters and tags (and icons or other attributes, potentially) as buttons that are toggled on/off when clicked. A clicked button would cause the map to show the results of applying the filter. However, it would also "filter" the panel buttons and leave visible only the buttons that would further reduce the number of visible nodes. For instance, if I clicked the tag "to-do" it would show only nodes with that tag, and the filter panel buttons would then only include buttons for tags that are attached to the currently visible nodes (that is, the nodes tagged "to-do".) Button "to-do" would remain visibly selected until clicked again. If "to-do" is clicked, and then I click tag "at office", the map would show only nodes with tags "to-do" and "office", and the panel would show only tag buttons for tags attached to nodes with both "to-do" and "office". This is a very fast way to combine filters and immediately see the results.

Clicking multiple buttons would act as an "and" filter combination. There would be no "or" option, but that is still available in the current filter composer (it's possible, of course, to make the buttons work with "or", but the filter button panel would not aim to be as complicated or complete as the current filter composer). Although the panel is less powerful in terms of possible options, it would be _much_ easier to use for 90% of the filtering I actually do.

This approach allows a way of using tags as filters to drill-down from all nodes to the subset of nodes I want to work with at that time. After I click one tag button, I can see immediately what other set of tags are also attached to the filtered nodes, and I can quickly filter out other nodes. That's something I can't do as easily with the current system, where I have to go to a different window that hides the map, create a filter for a specific tag, add it, apply it, see the results, go back to the compose window, create a 2nd filter, create a combined filter, save it if I want to use it later, apply it, go back to the map and see the results, etc.

I've talked mostly about tags as filters because that's how I imagine using them. I'm thinking that tags would be automatically added to this panel. But it would also be great to be able to add styles (filter to only show nodes with certain styles) or other attributes.

I really just want a simple way to filter my large map down to just the relevant nodes quickly and easily. I hope others who agree will post or suggest refinements to this idea.

**Comment by dpolivaev** [comment]
I do not see the buttons because the UI should support hundreds of tags.
But I think I could implement logic described at https://www.zotero.org/support/collections_and_tags#the_tag_selector using a drop-down box as they suggest.

**Reply by quickfold** [reply]
Thanks for considering this. I'm not sure what you mean by drop-down box. Zotero doesn't use a drop-down box to choose tags, just to display a menu with a few general options. I don't see how a drop-down box could provide the functionality I describe.

Zotero supports hundreds of tags, just as Freeplane might. It is not a problem for the Zotero user because the tag panel has a search box that filters the tags as you type. Type 3 letters and the hundreds of tags become 5 tags. That is one possible solution to making tags usable.

But even without such a search box, please note that the user who is interested in this filtering by button just needs to click one of the relevant tags and suddenly the hundreds of tags displayed gets reduced to less than a dozen (for me), since the UI only displays tags that share nodes with the selected tag. The panel is intended as more dynamic than the icon panel. The user doesn't need to see everything available, just enough to let them start the filtering process, and then the UI is easy to use.

**Comment by dpolivaev** [comment]
Well, let us think together. I can take a search field combined with drop-down box from the tag editor and put it into the tool panel. It shows all tags, that are filtered as you type. They can be additionally filtered depending on the presence of tags in nodes passing the current filter.

When you pick a tag, it can be either added as additional restrictive filter or added to each of currently selected nodes. Because for each case we need another set of available tags, it can be controlled by radio buttons.

I can display all tags added to the filter in an extra text field and remove them from the filter if you click on them.

I am not sure if displaying all tags attached to map nodes currently matching the filter is needed.

How do you see it?

Can you give me examples for use of the discussed functionality in the real life?

**Comment by dpolivaev** [comment]
I am considering adding a tag tab to the tool panel at the right side of Freeplane.

- It could contain a drop-down box used for adding tags to all selected nodes,
- It could contain a panel containing all tags available at currently selected nodes, and clicking on the tag in the box could be used for removing of the tag from the selected nodes,
- It could contain a filter section having a drop-down box for selecting a tag for filter and a panel showing all tags already used for filtering. The drop-down box could be limited to the tags available at nodes currently matching the filter and not already used for filtering.

I am not quite sure if the drop-down boxes and panels should display categorized or uncategorized tags.

Please share with me your ideas and use cases for such tools so that I can make better decisions.

**Comment by quickfold** [comment]
I haven't responded earlier because I was also thinking about the best way to move forward. I do not want to derail the discussion with a new topic, but in thinking about your suggestion I have realized that one usability issue in FP is that both the filter composer and the tag editor combine functions for editing tags & filters with functions for using them to manage mindmaps. But, an optimal UI for creating and editing may not be optimal for regular use, and this should be considered when thinking about how to implement a filter buttons panel.  For example, the filter composer and tag editor are not docked panels, so they block a significant amount of the map when it's open.  They are windows that have an "okay" button on them, implying they are not supposed to stay open while editing the map, but are just for setting things temporarily.  Compare this to the tool panel that is used to format notes.  Everyone is different, but I almost never use that tool panel, and I prefer to style nodes using the dedicated styles editor.  In contrast, if I am able to switch to using tags regularly, I imagine tagging notes quite frequently, and I would want to panel that is designed to stay open and to let me see my options by scanning it visually, rather than through a drop down or text search.

Let me explain one use case. I have a complicated mind map I used to plan my classes. It has all the information about the class, including all teaching materials like lectures and tests, and I often need to filter out particular subsets of information for certain uses, such as creating a syllabus or printing out lecture notes.  Because there are many categories of information, selecting the right combination of categories is quite complicated. Currently, I'm trying to use filters to create the right combinations, but it is very hard for me to predict the outcome of certain filters or combinations of filters when using the filter composer. For me, trying to compose the right filter is like trying to solve an abstract logic puzzle in my mind, and it is very hard for me to imagine the outcome of certain filter combinations. I have to think through it for a long time and build a filter and only then can I apply it  to see which notes it selects. Usually, I guess wrong and have to try again.  I am pretty confident that the filter composer feels very intuitive to you, as people's minds work quite differently. I would like a way to use buttons to build a filter in a more visual way, where pressing a button immediately shows me what notes the filter would select. And then I could build a more complicated filter by adding or removing additional elements with buttons. Perhaps there could be a way for me to save the filter once I found the right combination to select the notes I want. Does this make sense to you?

**Reply by euu2021** [reply]
> I have to think through it for a long time and build a filter and only then can I apply it to see which notes it selects. Usually, I guess wrong and have to try again. I am pretty confident that the filter composer feels very intuitive to you, as people's minds work quite differently. I would like a way to use buttons to build a filter in a more visual way, where pressing a button immediately shows me what notes the filter would select. And then I could build a more complicated filter by adding or removing additional elements with buttons. Perhaps there could be a way for me to save the filter once I found the right combination to select the notes I want.


Doesn't the filter composer has all of that, already?

> I would like a way to use buttons to build a filter in a more visual way, where pressing a button immediately shows me what notes the filter would select.

When you select an item in the filter composer, the nodes that match the filter get highlighted in the map.

> And then I could build a more complicated filter by adding or removing additional elements with buttons.

You can build the filter criteria little by little, by combining conditions.

> Perhaps there could be a way for me to save the filter once I found the right combination to select the notes I want.

The filter can be saved.

![javaw_1PlUgCmdRl](https://github.com/user-attachments/assets/5c889178-c5cf-4406-aa57-cf910bdd3cf3)

**Reply by quickfold** [reply]
Thanks for your comment. You are right, the filter composer can provide almost all of the functions that I mentioned, and I mentioned that for my first post. The issue is not whether the filter composer can do these things, but whether it does them in a way that is easy to use. In your video, you show that someone can combine filters in the way that I mentioned. What your video skips is the tedious process of populating the composer list with filters that you might want to combine. To add a filter to the list, you have to pick from a bunch of drop-down menus to create a combination of attributes and then manually add it to the list. And to combine filters, you have to do the same thing for each filter than you want to potentially combine with others. You have to think through it each time what you might want to combine, and why, and build a list of options one by one, and only then you can combine them, and if you want to try anything else, you have to add it to the list. I'm imagining something much easier to use, where the options are all laid out like in the icon pane. You can click them on and off easily, and they are all available at once.  You don't have to search your mind to think through what filter options are available or pull down a bunch of different drop-down menus to see what you might filter on. You just look at the available filters in a panel, visually scan them, and click on various things to try them.

**Reply by dpolivaev** [reply]
Actually, you can always add filter to nodes selected by current filter using "and filter"
<img width="436" alt="andFilter" src="https://github.com/user-attachments/assets/8e62cb0b-2047-4b7c-93c9-0bf9b314e72a">

**Reply by quickfold** [reply]
Yes, but once again, the issue is not whether there is a way to do something with filters with the current interface. The current interface is powerful but very hard to use. So the question is whether there is any interest in any change to make it easier to use. I explained my use case and I think I explained why the current interface is very hard for me (and likely many other people) to use.

**Comment by dpolivaev** [comment]
I significantly reworked representation of filter conditions in the filter toolbar.
Additionally, when you hover over the condition drop-down-box, Freeplane shows a tooltip where you can remove parts of conditions by mouse click.

<img width="724" alt="conditions" src="https://github.com/user-attachments/assets/dd6f6496-af86-43b1-b92a-7fd9c6b689bd">


It is work in progress, and I would like to get your feedback.

**Reply by euu2021** [reply]
This interactive tooltip is nice.
